python/nammu/test/test_nammu.py

```diff
@@ -243,3 +251,115 @@ def test_arabic_split_pane(self, monkeypatch):
 
         assert isinstance(self.nammu.atfAreaController.view.container,
                           JSplitPane)
+
+    def test_edit_compound(self, empty_compound):
+        '''
+        Adding a character should trigger several edit events (insert and
+        update colouring), added to the undo manager as a single edit compound
+        so they are all undone/redone at once.
+        '''
+        controller = self.nammu.atfAreaController
+        controller.edit_area.setText("a")
+        listener = controller.view.edit_listener
+        # Edits list within the undo_manager is not accessible and toString()
+        # is the only way I found to inspect it.
+        # TODO: I could do some of this stuff in Java, where I can access the
+        # edits, but that's a bit overcomplicated at this stage.
+        assert (listener.current_compound != empty_compound and
+                "edits: []" not in controller.undo_manager.toString())
+
+    def test_undo_empty_pane(self):
+        '''
+        Undo when empty pane should not do anything and the undo stackpile
+        should remain empty.
+        This test is needed because undoing empty panels is unstable and
+        sometimes raises exceptions.
+        '''
+        controller = self.nammu.atfAreaController
+        controller.clearAtfArea()
+        controller.undo()
+        assert ("edits: []" in controller.undo_manager.toString())
+
+    def test_redo_empty_pane(self):
+        '''
+        Redo when empty pane should not do anything and the undo stackpile
+        should remain empty.
+        This test is needed because undoing empty panels is unstable and
+        sometimes raises exceptions.
+        '''
+        controller = self.nammu.atfAreaController
+        controller.clearAtfArea()
+        controller.redo()
+        assert ("edits: []" in controller.undo_manager.toString())
+
+    def test_undo_after_opening_file(self, monkeypatch):
+        '''
+        Undo after closing a file should not bring back the old file.
+        '''
+        import javax.swing.JFileChooser
```

Comment:
This could go in the top-level imports (same in the next test)

Reply:
Done.